src/core/namespace.c

```diff
@@ -287,10 +288,18 @@ static int apply_mount(
 
                 /* First, get rid of everything that is below if there
                  * is anything... Then, overmount it with an
-                 * inaccessible directory. */
+                 * inaccessible path. */
                 umount_recursive(m->path, 0);
 
-                what = "/run/systemd/inaccessible";
+                // mount() always follow symlinks, so we follow symlinks during the stat() as well
+                r = stat(m->path, &target);
+                if (r != 0)
+                        return r;
+                if (!S_ISDIR(target.st_mode))
+                        what = "/dev/null";
+                else {
+                        what = "/run/systemd/inaccessible";
+                }
                 break;
```

Comment:
The config "/run/inaccessible/{blk,chr} rwm" will need to be added in:
- systemd source code:
 - https://github.com/systemd/systemd/blob/v230/src/nspawn/nspawn-register.c#L115
 - https://github.com/systemd/systemd/blob/v230/units/systemd-nspawn%40.service.in
- rkt documentation about starting rkt with a systemd unit file:
 - https://github.com/coreos/rkt/blob/master/Documentation/using-rkt-with-systemd.md#simple-unit-file
- In general, unit files written by users

Does it look correct?

> Other than that, I think as a special magic we could permit a syntax of "DeviceAllow=c0:0 rwm" and "DeviceAllow=b0:0 rwm" or so, the same way as udev uses it for places such as /run/udev/data/*.

For rkt, I would prefer to avoid adding new syntax for the DeviceAllow string:

rkt's stage1 currently includes systemd-nspawn v229. If we update rkt's stage1 to use a new systemd-nspawn with this PR (e.g. v231) and we run rkt on a host with systemd <= v230 / systemd-machined <= c230, will it work?

The "DeviceAllow" string is generated by systemd-nspawn (when started from the command line instead of a systemd service) and it is parsed by either systemd-machined or systemd on the host.